common/src/stack/command/stack/commands/load/json/__init__.py

```diff
@@ -0,0 +1,141 @@
+	data for each os.
+	</arg>
+
+	<arg optional='1' type='string' name='appliance'>
+	Load name, attr, route, firewall, partition, and controller
+	data for each appliance.
+	</arg>
+
+	<arg optional='1' type='string' name='group'>
+	Load name data for each group.
+	</arg>
+
+	<arg optional='1' type='string' name='bootaction'>
+	Load name, kernel, ramdisk, type, arg, and os data for each
+	bootaction.
+	</arg>
+	"""
+
+	def checks(self, args):
+
+		# Make sure that the user is not attepting to add more than one frontend
+		if ('host' in args or not args) and 'host' in self.data:
+			current_frontend = json.loads(self.command('list.host', [ 'a:frontend', 'output-format=json' ]))[0]['host']
+			for host in self.data['host']:
+				if host['appliance'] == 'frontend' and host['name'] != current_frontend:
+					raise CommandError(self, 'frontend name in the json document does not match the current frontend name')
+
+		# Make sure that all of the pallets at least have URLs
+		if ('software' in args or not args) and 'software' in self.data and 'pallet' in self.data['software']:
+			for pallet in self.data['software']['pallet']:
+				if pallet['url'] == None:
+					raise CommandError(self, f'pallet {pallet["name"]} {pallet["version"]} has no url')
+
+
+	def run(self, params, args):
+
+		filename, = self.fillParams([
+			('file', None, True),
+		])
+
+		if not os.path.exists(filename):
+			raise CommandError(self, f'file {filename} does not exist')
+
+		with open (os.path.abspath(filename), 'r') as f:
+			try:
+				self.data = json.load(f)
+			except ValueError:
+				raise CommandError(self, 'Invalid json document')
+
+		# run a few pre checks
+		self.checks(args)
+
+		# make a backup of the database in its current state in the event of any errors
+		s = subprocess.run(['mysqldump', 'cluster'], stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+		with open ('cluster_backup.sql', 'wb+') as f:
+			f.write(s.stdout)
+
+		# so that we are able to report how successful the load was
+		self.successes = 0
+		self.warnings = 0
+		self.errors = 0
+
+		# start a logger
+		self.log = logging.getLogger("load-json")
+		logging.basicConfig(filename='/var/log/load-json.log', filemode='w+', level=logging.INFO)
+
+		self.runPlugins(args)
+
+		# report how well the load went
+		self.notify(f'\nload finished with:\n{self.successes} successes\n{self.warnings} warnings\n{self.errors} errors\n')
+
+		# if there are errors, revert db changes and raise a CommandError
+		if self.errors != 0:
+			with open('cluster_backup.sql', 'rb') as f:
+				cluster_backup = f.read()
+			s = subprocess.Popen(['mysql', 'cluster'], stdin=subprocess.PIPE, stdout=subprocess.PIPE)
+			s.communicate(cluster_backup)
+
+			raise CommandError(self, 'There was at least one error, database has been reverted. No changes have been made.')
+		else:
+			# our load was successful so we can delete our temporary backup
+			s = subprocess.call(['rm', '-f', 'cluster_backup.sql'], stdout=subprocess.PIPE)
```

Comment:
If using the stacki scripts for creating backups, don't delete the backup, anymore